File: java/blackjack/src/main/java/Blackjack.java
import java.util.AbstractMap;
import java.util.Map;

public class Blackjack {
    private static final int BLACKJACK_VALUE = 21;
    private static final Map<String, Integer> CARD_VALUES = Map.ofEntries(
            new AbstractMap.SimpleEntry<>("ace", 11),
            new AbstractMap.SimpleEntry<>("two", 2),
            new AbstractMap.SimpleEntry<>("three", 3),
            new AbstractMap.SimpleEntry<>("four", 4),
            new AbstractMap.SimpleEntry<>("five", 5),
            new AbstractMap.SimpleEntry<>("six", 6),
            new AbstractMap.SimpleEntry<>("seven", 7),
            new AbstractMap.SimpleEntry<>("eight", 8),
            new AbstractMap.SimpleEntry<>("nine", 9),
            new AbstractMap.SimpleEntry<>("ten", 10),
            new AbstractMap.SimpleEntry<>("jack", 10),
            new AbstractMap.SimpleEntry<>("queen", 10),
            new AbstractMap.SimpleEntry<>("king", 10),
            new AbstractMap.SimpleEntry<>("other", 0)
    );
    public int parseCard(String card) {
        return CARD_VALUES.get(card);
    }

    public boolean isBlackjack(String card1, String card2) {
        return parseCard(card1) + parseCard(card2) == BLACKJACK_VALUE;
    }

    public String largeHand(boolean isBlackjack, int dealerScore) {
        if (!isBlackjack) return "P";
        if (dealerScore < 10) return "W";
        return "S";
    }

    public String smallHand(int handScore, int dealerScore) {
        if (handScore >= 17) return "S";
        if (handScore <= 11) return "H";
        if (dealerScore < 7) return "S";
        return "H";
    }

    // FirstTurn returns the semi-optimal decision for the first turn, given the cards of the player and the dealer.
    // This function is already implemented and does not need to be edited. It pulls the other functions together in a
    // complete decision tree for the first turn.
    public String firstTurn(String card1, String card2, String dealerCard) {
        int handScore = parseCard(card1) + parseCard(card2);
        int dealerScore = parseCard(dealerCard);

        if (20 < handScore) {
            return largeHand(isBlackjack(card1, card2), dealerScore);
        } else {
            return smallHand(handScore, dealerScore);
        }
    }
}

File: ruby/blackjack/blackjack.rb
# frozen_string_literal: true

module Blackjack
  def self.parse_card(card)
    case card
    when 'ace'
      11
    when 'two'
      2
    when 'three'
      3
    when 'four'
      4
    when 'five'
      5
    when 'six'
      6
    when 'seven'
      7
    when 'eight'
      8
    when 'nine'
      9
    when 'ten'
      10
    when 'jack'
      10
    when 'queen'
      10
    when 'king'
      10
    else
      0
    end
  end

  def self.card_range(card1, card2)
    case (parse_card card1) + (parse_card card2)
    when ..11
      'low'
    when ..16
      'mid'
    when ..20
      'high'
    else
      'blackjack'
    end
  end

  def self.first_turn(_card1, _card2, _dealer_card)
    raise 'Please implement the Blackjack.first_turn method'
  end
end
